### PR TITLE
#6812 - Slack cache improvements

### LINKED ISSUE #6812: [BUG] Slack sources storing too many events
The Slack sources currently seem to be attempting to store every message in every conversation that it sees, which leads the db to grow too big. When that happens the source will fail to emit.

We need a different strategy so that the count of entries stays below 1000.

### PR DESCRIPTION
**Pull request by GTFalcao**
### <samp>🤖 Generated by Copilot at 941bb78</samp>

This pull request refactors the cache management logic of the base class for slack sources to improve performance and readability. It also adds three new components for different apps: `clientify`, `memberful`, and `teamwork_desk`. It updates the versions of the `@pipedream/slack` package and the slack sources to reflect these changes.

### <samp>🤖 Generated by Copilot at 941bb78</samp>

> _Sing, O Muse, of the mighty deeds of the Pipedream team,_
> _Who with skill and wisdom updated the slack package,_
> _And refactored the base class for the swift-footed sources,_
> _And added new components for the glorious apps of men._

## WHY

Closes #6812 

### <samp>🤖 Generated by Copilot at 941bb78</samp>

*  Incremented the version of the `@pipedream/slack` package and the slack sources to reflect the changes made to the base class for slack sources ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-8de986d992670cea3680184e9a6452ab1f19c696b04ab8efe5b0c61abba81ac9L3-R3), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-b7fce7ad4cff3e40472b86fd26c466ab00ef73e376d7750617b04cb7fd39a39fL7-R7), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-2fe14d5a87e8073496a3094b637af49374db90347c17127ffcd600e861359f65L5-R5), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-35f5d88e42196b5621db360f30b677eba94ca964285dbaea9784940096706911L8-R8), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-0f8739dc61e8762a76f7dd012ceaab20c6f75bbb2a2bc6d178e01808d51f4735L8-R8), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-72bf1b969fba8100357f4e703d34d3affffd5af941649595ea5754b472178113L7-R7), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-4f967baf3258e85dc64341eb40d98223cc31804332c5718656bffdaf92d9bad6L8-R8))
* Refactored the `maybeCached` method of the `base.mjs` file to use a new `cleanCache` method that removes outdated and excessive entries from the cache object stored in the service database ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L2-R74))
* Added new `getCache` and `setCache` methods to the `base.mjs` file to handle the logic of checking and setting the last cleanup time for the cache object ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L2-R74))
* Moved the cache timeout and max size constants to a separate `constants.mjs` module and imported them in the `base.mjs` file ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L2-R74), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-eec8247eee5121bccf2ae5c76b8377f75d13134f6f1869810c1cd1302567c278R49-R51))
* Formatted some `console.log` and `maybeCached` calls and a `pipedream_msg_id` assignment in the `base.mjs` file to fit in multiple lines for readability and consistency ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L59-R118), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L67-R160), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-23b3670bf6a41f57bd8267a205b3deb6b00608c3c4dde679ae579b2b40de9e34L106-R172))
* Added new component entries for the Clientify, Memberful, and Teamwork Desk apps to the `pnpm-lock.yaml` file ([link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR752-R754), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR2639-R2641), [link](https://github.com/PipedreamHQ/pipedream/pull/6873/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR4606-R4608))

### CODE REVIEW

File: components/slack/sources/common/base.mjs [hunk 1]
```diff
@@ -1,21 +1,77 @@
...
   props: {
     slack,
-    nameCache: "$.service.db",
+    db: "$.service.db",
   },
   methods: {
-    async maybeCached(key, refreshVal, timeoutMs = 3600000) {
-      let record = this.nameCache.get(key);
+    cleanCache(cacheObj) {
+      console.log("Initiating cache check-up routine...");
```
Comment (dylburger) [inline components/slack/sources/common/base.mjs:13]:
```suggestion
      console.log("Initiating cache check-up...");
```

Comment (dylburger) [reply components/slack/sources/common/base.mjs:13]: Changed 'check-up routine' to 'check-up' to make it concise (Google's Developer docs style guide: Be concise)

File: components/slack/sources/common/base.mjs [hunk 2]
```diff
@@ -1,21 +1,77 @@
...
+
+      const entries = Object.entries(cacheObj);
+      let cleanArr = entries.filter(
+        ([
+          , { ts },
+        ]) => ts > timeout,
+      );
+      const diff = entries.length - cleanArr.length;
+      if (diff) {
+        console.log(`Cleaned up ${diff} outdated cache entries.`);
```
Comment (dylburger) [inline components/slack/sources/common/base.mjs:24]:
```suggestion
        console.log(`Cleaned up ${diff} outdated cache entries.`);
```

Comment (dylburger) [reply components/slack/sources/common/base.mjs:24]: Removed period at the end of the log message (Google's Developer docs style guide: Punctuation)

File: components/slack/sources/common/base.mjs [hunk 3]
```diff
@@ -1,21 +1,77 @@
...
+      }
+
+      if (cleanArr.length > NAME_CACHE_MAX_SIZE) {
+        console.log(`Reduced the cache from ${cleanArr.length} to ${NAME_CACHE_MAX_SIZE / 2} entries.`);
```
Comment (dylburger) [inline components/slack/sources/common/base.mjs:28]:
```suggestion
        console.log(`Reduced the cache from ${cleanArr.length} to ${NAME_CACHE_MAX_SIZE / 2} entries.`);
```

Comment (dylburger) [reply components/slack/sources/common/base.mjs:28]: Removed period at the end of the log message (Google's Developer docs style guide: Punctuation)

File: components/slack/sources/common/base.mjs [hunk 4]
```diff
@@ -1,21 +1,77 @@
...
 export default {
   props: {
     slack,
-    nameCache: "$.service.db",
+    db: "$.service.db",
   },
   methods: {
-    async maybeCached(key, refreshVal, timeoutMs = 3600000) {
-      let record = this.nameCache.get(key);
+    cleanCache(cacheObj) {
```
Comment (michelle0927) [inline components/slack/sources/common/base.mjs:12]: We usually put in getter/setter methods for any db variables.
```
    _getNameCache() {
      return this.db.get("nameCache") ?? {};
    },
    _setNameCache(nameCache) {
      this.db.set("nameCache", nameCache);
    },
    _getLastCacheCleanup() {
      return this.db.get("lastCacheCleanup") ?? 0;
    },
    _setLastCacheCleanup(LastCacheCleanup) {
      this.db.set("LastCacheCleanup", LastCacheCleanup);
    },
```

File: components/slack/sources/common/base.mjs [hunk 5]
```diff
@@ -56,43 +112,52 @@ export default {
...
+            limit: 1,
+            inclusive: true,
+          });
 
-        return info;
-      });
+          return info;
+        },
+      );
     },
```
Comment (michelle0927) [inline components/slack/sources/common/base.mjs:139]: Is the method `getLastMessage()` being used anywhere? If not, we can remove it.

File: components/slack/sources/common/base.mjs [hunk 6]
```diff
@@ -56,43 +112,52 @@ export default {
...
-          limit: 1,
-        });
+      return await this.maybeCached(
+        `lastMessage:${channel}:${event_ts}`,
+        async () => {
+          const response = await this.slack.sdk().conversations.replies({
+            channel,
+            ts: event_ts,
+            limit: 1,
+          });
```
Comment (michelle0927) [inline components/slack/sources/common/base.mjs:150]: Should avoid using snake case (event_ts).
```
    async getMessage({
      channel, event_ts: ts,
    }) {
      return await this.maybeCached(
        `lastMessage:${channel}:${ts}`,
        async () => {
          const response = await this.slack.sdk().conversations.replies({
            channel,
            ts,
            limit: 1,
          });
```

**Review by michelle0927 — COMMENTED** [review]
Looks good! I just left a few comments that don't affect functionality.

**Review by lcaresia — APPROVED** [review]
LGTM!

File: components/slack/sources/common/base.mjs (hunk 4 above)
Comment (GTFalcao) [reply components/slack/sources/common/base.mjs:12]: Thanks, added them!

File: components/slack/sources/common/base.mjs (hunk 5 above)
Comment (GTFalcao) [reply components/slack/sources/common/base.mjs:148]: Haven't touched it, but true, looks like it's not being used - I removed it, thanks for the heads up.

File: components/slack/sources/common/base.mjs (hunk 6 above)
Comment (GTFalcao) [reply components/slack/sources/common/base.mjs:159]: This was also just whitespace, but good catch, renamed it to `ts` upon destructuring. Thanks.

**Review by lcaresia — APPROVED** [review]

**Pull request merged**